mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java

```diff
@@ -19,11 +19,33 @@
 
 package org.apache.james.mailbox.events.delivery;
 
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+
 import org.apache.james.mailbox.Event;
 import org.apache.james.mailbox.MailboxListener;
 
 public interface EventDelivery {
+    class ExecutionStages {
+        private final CompletableFuture<Void> synchronousListenerFuture;
+        private final CompletableFuture<Void> asynchronousListenerFuture;
+
+        ExecutionStages(CompletableFuture<Void> synchronousListenerFuture, CompletableFuture<Void> asynchronousListenerFuture) {
+            this.synchronousListenerFuture = synchronousListenerFuture;
```

Comment:
why don't you expose Mono instead ?

Reply:
The big deal is that with the mono the computation did not started yet.

I don't believe the caller should be responsible of starting these computations....

Reply:
just call `.cache()` then

Reply:
Yes `.cache()` combined with `.subscribe()` for triggering the initial computation...

Reply:
No, cache is enough, it starts the computation

Reply:
I do believe you but I also do have some tests failures if I do not call `.subscribe()`...

Reply:
Ok at least it's what the doc says : https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Mono.html#cache--

But looking at the picture, it looks like the doc is wrong, trust your test :+1: